First draft of embedding blog post (#426)
readthedocs/ethicalads.io: content/posts/2024-niche-ad-targeting.md

Title: Niche Targeting: Using AI to build better contextual ad targeting
Date: Apr 2, 2024
description: How we build better contextual ad targeting using embeddings.
tags: content-targeting, engineering
authors: Eric Holscher
image: /images/posts/niche-targeting.png

This is an update to our original post on [content-based ad targeting](https://www.ethicalads.io/blog/2022/11/a-new-approach-to-content-based-targeting-for-advertising/).
In this post, I'll talk a bit more about our next step, using AI (embeddings specifically) to build better contextual ad targeting.

At the end of our last post,
we were able to bucket pages into _Topics_ based on page content.
We did this by training a model with ~100 examples of each topic,

> and then running that model against the pages on our network.

Reviewer: This may be worth expanding before this blog goes live. This glosses over a lot of context about how we crawl sites on our ad network so we can learn about pages and later target them with ads. For somebody who hasn't read the previous post, that might be unclear.

This gave us a good starting point for targeting ads to pages,
but we wanted to get more granular.

## Targeting each page individually with embeddings

> Our new approach is to use [word embeddings](https://en.wikipedia.org/wiki/Word_embedding) to represent both the ads and the publisher pages.

Reviewer: Should we mention that we're getting embeddings for an ad from advertiser landing/product pages?

This allows us to generate a representation of a document (like a page or an ad),
which can be compared against each other.

We're currently using the [Sentence Transformers](https://www.sbert.net/) library to generate these embeddings.
We will likely upgrade to a more advanced model in the future,
but this was perfect for our initial tests.

### Generating, storing, and querying embeddings

A simple example of what this looks like might be:

```python
# Generate embeddings for a page
import requests
from bs4 import BeautifulSoup
from sentence_transformers import SentenceTransformer

model = SentenceTransformer(MODEL_NAME, cache_folder=CACHE_FOLDER)
text = BeautifulSoup(requests.get(url), 'html.parser').get_text()
embedding = model.encode(text)

> return embedding.tolist()

Reviewer: ```suggestion
print(embedding.tolist())
```

```

We're then using [pgvector](https://github.com/pgvector/pgvector) and [pgvector-python](https://github.com/pgvector/pgvector-python) to manage these embeddings in Django & Postgres,
which is what we're already using in production.

```python
from django.db import models
from pgvector.django import VectorField

# Store the content in Postgres/Django

class Embedding(models.Model):

# FK where we keep metadata about the URL
analyzed_url = models.ForeignKey(
    AnalyzedUrl,
    on_delete=models.CASCADE,
    related_name="embeddings",
)

# Model name so we can use different models in the future
model = models.CharField(max_length=255, default=None, null=True, blank=True)

# The actual embedding
vector = VectorField(dimensions=384, default=None, null=True, blank=True)
```

> Then we're able to query the database for the most similar pages to a given ad:

Reviewer: To the advertiser's landing page?

* We were able to implement this approach with minimal changes to our existing infrastructure, mostly because we're already heavily invested in the Python ecosystem and Postgres.

## Challenges and Considerations

We have a few challenges to overcome with this approach:

* We currently aim for 50ms for ad response time, and this approach is currently slower than that. We're working on optimizing this with indexing, and might look at using an in-memory vector store in the future.
* Embeddings currently work pretty well, but can often associate things that are not relevant. For example, we're run into issues where the same words are used to mean different things (eg. "model"), and the embeddings can get confused.

## Conclusion

This approach is still in its early stages, but we're excited about the potential.
The better we can get at ethical ad targeting,
everyone in our network benefits:

* **Advertisers** get better ad targeting, ensuring they show up in front of the right users.
* **Publishers** get more money while showing a single, minimalist ad.

> * **Users** get a better experience, with fewer ads that are more relevant to them.

Reviewer: This isn't fewer compared with our existing approach

Author: Sure, but the goal with higher CPMs is that publishers don't need as many ads. We're also selling our overall approach, not just the changes from this specific update.

Author: I updated this to not mention fewer ads.

This is our vision for advertising,
and we're excited about the potential of this approach.

> 
> ## Resources

Reviewer: ```suggestion
```

You can probably just remove this heading and just thank Simon at the end of the conclusion.

Title: Niche Targeting: Using AI to build better contextual ad targeting
Date: Apr 2, 2024

> description: How we build better contextual ad targeting using embeddings.

Reviewer: I think this description is a little short and a little weak.

```suggestion
description: How we build better contextual ad targeting using PostgreSQL, Django, and embeddings which are a way to encode the "relatedness" of text and pages for use in machine learning.
```

tags: content-targeting, engineering
authors: Eric Holscher

> image: /images/posts/niche-targeting.png

Reviewer: The double header in the image where we have our header plus the header inside the image is a little weird. Maybe this is unavoidable but maybe we could improve this by putting this screenshot inside the outline of a computer or something like that.

![Screenshot from 2024-03-27 14-09-23](https://github.com/readthedocs/ethicalads.io/assets/185043/11e24d55-8031-422f-b0aa-82f6c309fbb6)

Author: Yea, I'm gonna futz more with the screenshot. Probably just removing the page chrome, I assume.

This is an update to our original post on [content-based ad targeting](https://www.ethicalads.io/blog/2022/11/a-new-approach-to-content-based-targeting-for-advertising/).

> In this post, I'll talk a bit more about our next step, using AI (embeddings specifically) to build better contextual ad targeting.

Reviewer: ```suggestion
In this post, I'll talk a bit more about our next step, using machine learning (embeddings specifically) to build better contextual ad targeting.
```

AI is for the "marketing focused" post. If it's in Python, it's machine learning. If it's in powerpoint, it's AI.

Title: Using embeddings in production with Postgres & Django for niche ad targeting

> Date: Apr 2, 2024

Reviewer: ```suggestion
Date: Apr 23, 2024
```

tags: content-targeting, engineering
authors: Eric Holscher
image: /images/posts/niche-targeting.png

This is an update to our original post on [content-based ad targeting](https://www.ethicalads.io/blog/2022/11/a-new-approach-to-content-based-targeting-for-advertising/).
In this post, I'll talk a bit more about our next step, using machine learning (embeddings specifically) to build better contextual ad targeting.

At the end of our last post,
we were crawling all our publisher's pages,
and categorizing pages into _Topics_ based on page text.
We did this by training a model with ~100 examples of each topic,
and then storing the topics in our database for fast ad serving.

This gave us a good starting point for targeting ads by topic,
but we wanted to get more granular.

## Targeting each page individually with embeddings

Our new approach is to use [word embeddings](https://en.wikipedia.org/wiki/Word_embedding) to represent both the advertisers landing page and the publisher pages.
This allows us to generate a representation of these pages,
which can be compared against each other.

> We're currently using the [Sentence Transformers](https://www.sbert.net/) library to generate these embeddings.

Reviewer: ```suggestion
We're currently using Python's [SentenceTransformers](https://www.sbert.net/) library to generate these embeddings.
```

We will likely upgrade to a more advanced model in the future,
but this was perfect for our initial tests.

### Generating, storing, and querying embeddings

A simple example of what this looks like might be:

```python
import requests
from bs4 import BeautifulSoup
from sentence_transformers import SentenceTransformer

# Generate embeddings for a page

model = SentenceTransformer(MODEL_NAME, cache_folder=CACHE_FOLDER)
text = BeautifulSoup(requests.get(url), 'html.parser').get_text()
embedding = model.encode(text)
print(embedding.tolist())
```

We're then using [pgvector](https://github.com/pgvector/pgvector) and [pgvector-python](https://github.com/pgvector/pgvector-python) to manage these embeddings in Django & Postgres,

> which is what we're already using in production.

Reviewer: ```suggestion
which is what we're already using in production for serving ads.
```

Author: We're using it in all our production stack, so not really sure this makes sense.

```

Then we're able to query the database for the most similar publisher pages to an advertiser's landing page:

```python
from pgvector.django import CosineDistance
from .models import Embedding

# Find the most similar ads for the page we're serving an ad on

Embedding.objects.annotate(
    distance=CosineDistance("vector", embedding)
).order_by("distance")

```

## Try out a demo

You can see a screenshot of our niche targeting in action at the top of this page.
This is a simple proof of concept,
but you can see how we're able to target ads specifically focusing on MongoDB and Databases,
when serving a MongoDB ad.

You can [try out our Niche Targeting Demo](https://www.ethicalads.io/advertisers/similar-pages/?url=https%3A%2F%2Fwww.mongodb.com%2Fatlas),
and let us know how it goes!

## Advantages of Niche Targeting

There is a huge win both in terms of privacy and user experience with this approach:

* **We're able to target ads to pages without needing to know anything about the user.** The better we get at targeting, the more powerful our ethical advertising approach becomes, and the larger we can scale out network.

> * The user experience of minimalist, well-targeted ads is better. We're able to show fewer ads and charge more for them, which is a win-win for everyone.

Reviewer: ```suggestion
* The user experience of minimalist, well-targeted ads is better. We're able to show fewer ads and charge more for them because they perform better. This is a win-win for everyone.
```

* We were able to implement this approach with minimal changes to our existing infrastructure, mostly because we're already heavily invested in the Python ecosystem and Postgres.

## Challenges and Considerations

We have a few challenges to overcome with this approach:

* We currently aim for 50ms for ad response time, and this approach is currently slower than that. We're working on optimizing this with indexing, and might look at using an in-memory vector store in the future.
* Embeddings currently work pretty well, but can often associate things that are not relevant. For example, we're run into issues where the same words are used to mean different things (eg. "model"), and the embeddings can get confused.

## Conclusion

This approach is still in its early stages, but we're excited about the potential.
The better we can get at ethical ad targeting,
everyone in our network benefits:

* **Advertisers** get better ad targeting, ensuring they show up in front of the right users.

> * **Publishers** get more money while showing a single, minimalist ad.

Reviewer: ```suggestion
* **Publishers** get more money while showing a single ad rather than resorting to multiple, larger ads that take over their site
```
